### ISSUE TITLE
Bluetooth Kernel Bug: After connecting either HFP/HSP or A2DP is not available (Regression in 6.9.3, 6.8.12)

### ISSUE THREAD

**Issue by timoschroeder1213** [bluez/bluez#865]
Hallo, as requested by @Vudentz I open up this issue here.

Here a copy from my text I sent to Linux Kernel Mailing lists:

> Hallo,
> on my two notebooks, one with Ubuntu (Mainline Kernel 6.9.3, bluez
> 5.7.2) and the other one with Manjaro (6.9.3, bluez 5.7.6) I'm having
> problems with my Sony WH-1000XM3 and Shure BT1. Either A2DP or HFP/HSP
> is not available after the connection has been established after a
> reboot or a reconnection. It's reproducible that with the WH-1000XM3
> the A2DP profiles are missing and with the Shure BT1 HFP/HSP profiles
> are missing. It also takes longer than usual to connect and I have a
> log message in the journal:
> 
> Jun 06 16:28:10 liebig bluetoothd[854]:
> profiles/audio/avdtp.c:cancel_request() Discover: Connection timed out
> (110)
> 
> When I disable and re-enable bluetooth (while the Headsets are still
> on) and trigger a reconnect from the notebooks, A2DP and HFP/HSP
> Profiles are available again.
> 
> I also tested it with 6.8.12 and it's the same problem. 6.8.11 and
> 6.9.2 don't have the problem.
> So I did a bisection. After reverting commit
> af1d425b6dc67cd67809f835dd7afb6be4d43e03 "Bluetooth: HCI: Remove
> HCI_AMP support" for 6.9.3 it's working again without problems.
> 
> Let me know if you need anything from me.
> 
> Best regards,
> Timo
> 

I also built a kernel from current bluetooth-next and the problem also shows on both systems. When I revert commit 84a4bb6548a29326564f0e659fb8064503ecc1c7 the problem is gone. Additionally the problem is reproducible on two more systems. Maybe it's somewhat Headset related.

My systems are using pipewire and wireplumber. I attached inxi outputs from both systems, btmon traces and bluetoothd logs from the Manjaro system. I had first connected the headsets, then I disabled bluetooth, enabled it and reinitiated the connection from the notebook.

[bluetoothd_Shure-BT1.txt](https://github.com/user-attachments/files/15748003/bluetoothd_Shure-BT1.txt)
[bluetoothd_WH-1000XM3.txt](https://github.com/user-attachments/files/15748004/bluetoothd_WH-1000XM3.txt)
[btmon_Shure-BT1.txt](https://github.com/user-attachments/files/15748005/btmon_Shure-BT1.txt)
[btmon_WH-1000XM3.txt](https://github.com/user-attachments/files/15748006/btmon_WH-1000XM3.txt)
[inxi_lifebookE754_ubuntu.txt](https://github.com/user-attachments/files/15748007/inxi_lifebookE754_ubuntu.txt)
[inxi_thinkpadP14v_manjaro.txt](https://github.com/user-attachments/files/15748008/inxi_thinkpadP14v_manjaro.txt)

**Comment by Vudentz** [comment]
```
< ACL Data TX: Handle 4 flags 0x00 dlen 6                                                                                                                                                                                                                               #148 [hci0] 20.604660
      Channel: 387 len 2 [PSM 25 mode Basic (0x00)] {chan 2}
      AVDTP: Discover (0x01) Command (0x00) type 0x00 label 13 nosp 0
...
= bluetoothd: profiles/audio/avdtp.c:cancel_request() Discover: Connection timed out (110)                                                                                                                                                                                          26.561545
< ACL Data TX: Handle 4 flags 0x00 dlen 12                                                                                                                                                                                                                              #170 [hci0] 26.561927
      L2CAP: Disconnection Request (0x06) ident 12 len 4
        Destination CID: 387
        Source CID: 66
```

So the XM3 at least don't seem to be responding to AVDTP_Discover which is why we drop the connection to A2DP.

**Comment by Vudentz** [comment]
Also the fact that is not always A2DP that is missing sort of tell me this is not really specific to AVDTP_Discover though, so I wonder if the controller hasn't seem the message, but at least btmon shows that it has been transmitted in next frame since we got Number of Completed Packets.

The timeout is 6 seconds, which basically explain the slowdown, and there is at least 3 seconds of inactivity so it is not that it was busy the whole time.

**Comment by kode54** [comment]
This issue may be why my WH-1000XM4 was also failing to connect if I added wireplumber settings to force A2DP-only. A2DP managed to connect properly if I removed that override and allowed the defaults, but then it would manage to switch to HSP/HFP if anything started capturing audio, even with a different default device set, if some application probed info on all capture devices.

**Comment by pv** [comment]
Two similar reports on Pipewire issue tracker so far: broken on 6.9.3, AVDTP: Discover fails as above (packet completion event seen but no reply RX), Sony WH-1000XM3 and WH-XB900N. One report also on Mediatek controller so 3+ different controller manufacturers.

**Comment by kode54** [comment]
Re: A2DP-only, I was using this wireplumber snippet: https://wiki.archlinux.org/title/Bluetooth_headset#Disable_PipeWire_HSP/HFP_profile

```
monitor.bluez.rules = [
  {
    matches = [
      {
        ## This matches all bluetooth devices.
        device.name = "~bluez_card.*"
      }
    ]
    actions = {
      update-props = {
        bluez5.auto-connect = [ a2dp_sink ]
        bluez5.hw-volume = [ a2dp_sink ]
      }
    }
  }
]

monitor.bluez.properties = {
  bluez5.roles = [ a2dp_sink ]
  bluez5.hfphsp-backend = "none"
}
```

**Comment by Vudentz** [comment]
Looks like this was the culprit, weird that CI didn't catch it: https://patchwork.kernel.org/project/bluetooth/list/?series=860197

**Comment by pv** [comment]
Hmm, too bad, and increasingly mysterious... At the moment I don't see anything else suspicious in the commit.

I haven't been able to reproduce this, but I don't have one of the known-affected headsets. Can you specify a bit more how to reproduce it, e.g. only after boot or after `bluetoothctl disconnect`?

If you could attach a btmon trace that appears with 84a4bb6548a2 reverted, maybe there is something is different in the connection setup sequence in the two cases.

**Comment by timoschroeder1213** [comment]
After a fresh reboot I have the problem when I switch on my WH-1000XM3 and it connects. After every further off and on switching the problem always shows. What helps is disconnecting the WH-1000XM3 by using Gnomes Quick Settings and then reinitiate the connects also through Gnomes Quick Settings. The (in the first post) mentioned disabling of bluetooth isn't necessary as I just learned when I was playing arround.
I attached traces and logs showing the first connect and the reconnect with an affected kernel.
I also attached traces and logs with a kernel were I reverted the causing commit.
[bluetoothd_WH-1000XM3_commit_reverted.txt](https://github.com/user-attachments/files/15752859/bluetoothd_WH-1000XM3_commit_reverted.txt)
[bluetoothd_WH-1000XM3_with_reconnect.txt](https://github.com/user-attachments/files/15752860/bluetoothd_WH-1000XM3_with_reconnect.txt)
[btmon_WH-1000XM3_commit_reverted.txt](https://github.com/user-attachments/files/15752861/btmon_WH-1000XM3_commit_reverted.txt)
[btmon_WH-1000XM3_with_reconnect.txt](https://github.com/user-attachments/files/15752862/btmon_WH-1000XM3_with_reconnect.txt)

**Comment by pv** [comment]
Ok, I see now my patch was wrong, I'll send a v2...

**Comment by timoschroeder1213** [comment]
The latest patch fixes the issue. Thank you!

**Comment by Vudentz** [comment]
Fixed by https://github.com/bluez/bluetooth-next/commit/9b70755997d372a80928faf90fd5ea4a637b90ac

**Comment by jirislaby** [comment]
> Fixed by [bluez/bluetooth-next@9b70755](https://github.com/bluez/bluetooth-next/commit/9b70755997d372a80928faf90fd5ea4a637b90ac)

Not 5fe0be8868aaa96fdd31bab5cfef079bf89a3fa2?

**Comment by jirislaby** [comment]
> Same change, just rebased.

Not at all, AFAICT. The former is "Bluetooth: hci_event: Fix setting of unicast qos interval" to fix #823.

**Comment by Vudentz** [comment]
> > Same change, just rebased.
> 
> Not at all, AFAICT. The former is "Bluetooth: hci_event: Fix setting of unicast qos interval" to fix #823.

Facepalm, right my bad, was probably not paying attention or was thinking this was a different problem, anyway good that you had commented on it.

**Comment by perrin4869** [comment]
thanks for the fix! I just tried with the latest 6.9.5, but it seems like it hasn't landed there yet
since I am not familiar with how `bluetooth-next` gets merged into the main tree, I was wondering in which kernel this fix is expected to land?

**Comment by Vudentz** [comment]
It should be on its way:

https://patchwork.kernel.org/project/bluetooth/patch/20240610135803.920662-1-luiz.dentz@gmail.com/

**Comment by gcem** [comment]
It's in 6.9.6:
https://cdn.kernel.org/pub/linux/kernel/v6.x/ChangeLog-6.9.6

**Comment by perrin4869** [comment]
can confirm the problem is gone in `6.9.6`
